File: app/src/main/kotlin/rpcnode/toolkit/servers/domain/model/AgentVersion.kt
package rpcnode.toolkit.servers.domain.model

/** Semver-ish: local older than remote. Equal or empty → not outdated. */
fun agentVersionOutdated(localRaw: String, remoteRaw: String): Boolean
{
    val local = stripV(localRaw)
    val remote = stripV(remoteRaw)
    if (local.isEmpty() || remote.isEmpty() || local == remote)
    {
        return false
    }
    val left = local.split('.')
    val right = remote.split('.')
    val n = maxOf(left.size, right.size)
    for (i in 0 until n)
    {
        val a = left.getOrNull(i)?.toIntOrNull() ?: 0
        val b = right.getOrNull(i)?.toIntOrNull() ?: 0
        if (a < b)
        {
            return true
        }
        if (a > b)
        {
            return false
        }
    }
    return false
}

private fun stripV(raw: String): String
{
    val value = raw.trim()
    if (value.length > 1 && (value[0] == 'v' || value[0] == 'V'))
    {
        return value.substring(1)
    }
    return value
}

File: admin/src/lib/agentVersion.ts
/** Semver-ish: local older than remote (same rules as panel agentVersionOutdated). */
export function agentVersionOutdated(local: string, remote: string): boolean {
  const a = local.trim().replace(/^v/i, '')
  const b = remote.trim().replace(/^v/i, '')
  if (!a || !b || a === b) return false
  const as = a.split('.')
  const bs = b.split('.')
  const n = Math.max(as.length, bs.length)
  for (let i = 0; i < n; i++) {
    const ai = parseInt(as[i] || '0', 10) || 0
    const bi = parseInt(bs[i] || '0', 10) || 0
    if (ai < bi) return true
    if (ai > bi) return false
  }
  return false
}
